### ISSUE TITLE
 Object Detection Grocery Tutorial: BB = BB[sorted_ind, :] IndexError: too many indices for array

### ISSUE THREAD

**Issue by StuartRucker** [Microsoft/CNTK#2039]
I am following the Fast R-Cnn Grocery [Object Detection tutorial](https://docs.microsoft.com/en-us/cognitive-toolkit/Object-Detection-using-Fast-R-CNN), and all of the scripts work up to `A3_ParseAndEvaluateOutput.py`. Then, I am faced with the following error:

```
File "C:\Users\stuart.rucker\Documents\CNTK-master\Examples\Image\Detection\FastRCNN\imdb_data.py", line 258, in _voc_computePrecisionRecallAp
    BB = BB[sorted_ind, :]
IndexError: too many indices for array
```

I did a little looking through the code and it seems like the issue might begin on line 319 of test.py 
 when `all_boxes` is emptied, but really, I have no idea.

```
       for j in range(1, imdb.num_classes):
            for i in range(num_images):
                inds = np.where(all_boxes[j][i][:, -1] > thresh[j])[0]
                if len(inds) == 0:
                    all_boxes[j][i] = []
                else:
                    all_boxes[j][i] = all_boxes[j][i][inds, :]
```

**Comment by SandyaSivakumar** [comment]
@StuartRucker  - I am also facing the same problem, please do let me know if you have found the solution. Thanks in advance.

**Comment by pkranen** [comment]
Which CNTK version are you using? I tried just now with the latest master and A1, A2 and A3 work fine.

**Comment by SandyaSivakumar** [comment]
I have used CNTK version 2.0.
I have not yet tried the updated master. Will try and let you know the results.

**Comment by SandyaSivakumar** [comment]
@pkranen  - Sir, which classifier are you using,  'svm' or 'nn' ??

**Comment by pkranen** [comment]
From our GitHub repository we only use the 'nn' classifier. For details on svm in combination with CNTK Fast R-CNN please see https://github.com/Azure/ObjectDetectionUsingCntk#step-4-classifier-training.

**Comment by kyoro1** [comment]
After refreshing repository in CNTK and executing `python A2_RunWithPyModel.py`, I received the following error:

```sh
$ python A2_RunWithPyModel.py
--------------------------------------------------------------
2017-07-24 10:13:33
PARAMETERS: datasetName = Grocery
PARAMETERS: cntk_nrRois = 100
Traceback (most recent call last):
  File "A2_RunWithPyModel.py", line 213, in <module>
    trained_model = train_fast_rcnn()
  File "A2_RunWithPyModel.py", line 145, in train_fast_rcnn
    frcn_output = frcn_predictor(image_input, roi_input, num_classes, model_path)
  File "A2_RunWithPyModel.py", line 113, in frcn_predictor
    roi_out   = roipooling(conv_out, rois, C.MAX_POOLING, (roi_dim, roi_dim), 0.0625)
  File "C:\Anaconda\envs\py35\lib\site-packages\cntk\internal\swig_helper.py", line 69, in wrapper
    result = f(*args, **kwds)
TypeError: roipooling() takes from 3 to 4 positional arguments but 5 were given
```

So, I intentionally rollbacked the method `roipooling` as follows:
```py
## current 
    roi_out   = roipooling(conv_out, rois, C.MAX_POOLING, (roi_dim, roi_dim), 0.0625)
## What I used:
    roi_out   = roipooling(conv_out, rois, (roi_dim, roi_dim))
```

Here's a result

```sh
$ python A2_RunWithPyModel.py
--------------------------------------------------------------
2017-07-24 10:18:02
PARAMETERS: datasetName = Grocery
PARAMETERS: cntk_nrRois = 100
Training Fast R-CNN model for 20 epochs.
Training 54603793 parameters in 6 parameter tensors.
Learning rate per sample: 1e-05
Momentum per sample: 0.9048374180359595
Finished Epoch[1 of 20]: [Training] loss = 278.214941 * 25, metric = 51.96% * 25 4.825s (  5.2 samples/s);
Finished Epoch[2 of 20]: [Training] loss = 260.710605 * 25, metric = 23.20% * 25 1.059s ( 23.6 samples/s);
Finished Epoch[3 of 20]: [Training] loss = 240.180313 * 25, metric = 23.20% * 25 1.018s ( 24.6 samples/s);
Finished Epoch[4 of 20]: [Training] loss = 212.694648 * 25, metric = 23.20% * 25 1.113s ( 22.5 samples/s);
Finished Epoch[5 of 20]: [Training] loss = 177.876484 * 25, metric = 23.20% * 25 1.154s ( 21.7 samples/s);
Finished Epoch[6 of 20]: [Training] loss = 144.604277 * 25, metric = 23.20% * 25 1.061s ( 23.6 samples/s);
Finished Epoch[7 of 20]: [Training] loss = 125.566191 * 25, metric = 23.20% * 25 0.995s ( 25.1 samples/s);
Finished Epoch[8 of 20]: [Training] loss = 119.546172 * 25, metric = 23.20% * 25 1.145s ( 21.8 samples/s);
Finished Epoch[9 of 20]: [Training] loss = 117.817344 * 25, metric = 23.20% * 25 1.055s ( 23.7 samples/s);
Finished Epoch[10 of 20]: [Training] loss = 117.436025 * 25, metric = 23.20% * 25 1.158s ( 21.6 samples/s);
Learning rate per sample: 1e-06
Finished Epoch[11 of 20]: [Training] loss = 117.250840 * 25, metric = 23.20% * 25 1.033s ( 24.2 samples/s);
Finished Epoch[12 of 20]: [Training] loss = 117.163604 * 25, metric = 23.20% * 25 1.095s ( 22.8 samples/s);
Finished Epoch[13 of 20]: [Training] loss = 117.148115 * 25, metric = 23.20% * 25 1.023s ( 24.4 samples/s);
Finished Epoch[14 of 20]: [Training] loss = 117.128389 * 25, metric = 23.20% * 25 1.235s ( 20.2 samples/s);
Finished Epoch[15 of 20]: [Training] loss = 117.114023 * 25, metric = 23.20% * 25 1.097s ( 22.8 samples/s);
Learning rate per sample: 1e-07
Finished Epoch[16 of 20]: [Training] loss = 117.094053 * 25, metric = 23.20% * 25 1.195s ( 20.9 samples/s);
Finished Epoch[17 of 20]: [Training] loss = 117.089521 * 25, metric = 23.20% * 25 1.086s ( 23.0 samples/s);
Finished Epoch[18 of 20]: [Training] loss = 117.087676 * 25, metric = 23.20% * 25 1.074s ( 23.3 samples/s);
Finished Epoch[19 of 20]: [Training] loss = 117.086211 * 25, metric = 23.20% * 25 1.103s ( 22.7 samples/s);
Finished Epoch[20 of 20]: [Training] loss = 117.085244 * 25, metric = 23.20% * 25 1.017s ( 24.6 samples/s);
Stored trained model at C:\Users\kyiwasak\git\CNTK\Examples\Image\Detection\FastRCNN\Output\frcn_py.model
Evaluating Fast R-CNN model for 5 images.
-------------------------------------------------------------------
Build info:

                Built time: Apr 21 2017 01:32:20
                Last modified date: Fri Apr 21 00:23:15 2017
                Build type: Release
                Build target: GPU
                With 1bit-SGD: no
                With ASGD: yes
                Math lib: mkl
                CUDA_PATH: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v8.0
                CUB_PATH: c:\local\cub-1.4.1
                CUDNN_PATH: C:\local\cudnn-8.0-windows10-x64-v5.1
                Build Branch: HEAD
                Build SHA1: f190457aecbfe07c3fe27c89f36108bc2488727b
                Built by svcphil on liana-08-w
                Build Path: C:\jenkins\workspace\CNTK-Build-Windows\Source\CNTKv2LibraryDll\
                MPI distribution: Microsoft MPI
                MPI version: 7.0.12437.6
-------------------------------------------------------------------
```

As you see, the convergence of `loss` is very slow. 

Also, I received the same error as the top of this issue.

@pkranen , Did your trial similar to the tutorial result in convergence?

**Comment by pkranen** [comment]
Closing this one since the user reported that the issue was fixed with the 2.1 release.

**Comment by pavanteja295** [comment]
I'm referring here to https://github.com/rbgirshick/py-faster-rcnn/ where the same error arises and the reason and the solution for the error is this:

I have seen a lot of posts regarding the error BB = BB[sorted_ind, :] IndexError: too many indices for array which i as well encountered while running the evaluation on a very weakly trained model. I actually found a bug in the file pascal_voc.py which causes this issue.
The all_boxes parameter for the py-faster-rcnn/lib/datasets/pascal_voc.py/evaluate_detection() function in pascal_voc.py is a list of numpy arrays. But in pascal_voc.py/_write_voc_results_file(all_boxes) function in line 251 we check if the

dets == []
but here dets is of type <class 'numpy.ndarray'> and the statement in line 251 evaluates to False leading to consider the empty array being written into the output detection file.

This file with some empty arrays when read in py-faster-rcnn/lib/datasets/eval_voc.py/voc_eval() in line 148 throws and error as BB is just an array.

Solution : In pascal_voc.py/_write_voc_results_file(all_boxes) edit to line 251 should be

dets.tolist() == []

This wont throw the error any more !

Kudos !
